File: Acquisition/LiveDisplay/VB.NET/Form1.vb
'*******************************************************************************
' Copyright (C) 2004-2010 Cognex Corporation
'
' Subject to Cognex Corporation's terms and conditions and license agreement,
' you are authorized to use and modify this source code in any way you find
' useful, provided the Software and/or the modified Software is used solely in
' conjunction with a Cognex Machine Vision System.  Furthermore you acknowledge
' and agree that Cognex has no warranty, obligations or liability for your use
' of the Software.
'*******************************************************************************
'
' This sample program is designed to illustrate certain VisionPro features or 
' techniques in the simplest way possible. It is not intended as the framework 
' for a complete application. In particular, the sample program may not provide
' proper error handling, event handling, cleanup, repeatability, and other 
' mechanisms that a commercial quality application requires.

' This sample demonstates how to start and stop a live display using the CogDisplay.
'
' This program assumes that you have some knowledge of Visual Basic and VisionPro
' programming.
'
' Follow the next steps in order to start and stop the live display
' Step 1) Create a  CogAcqFifoTool.
' Step 2) Stop the live display
' Step 3) Start the live display.
Option Explicit On 

Imports System
' Needed for VisionPro
Imports Cognex.VisionPro
' Needed for VisionPro Exceptions
Imports Cognex.VisionPro.Exceptions

Namespace LiveDisplay
  Public Class Form1
    Inherits System.Windows.Forms.Form
    Private WithEvents mTool As CogAcqFifoTool
    Private mAcqFifo As Cognex.VisionPro.ICogAcqFifo
        Private mBrightness As Cognex.VisionPro.ICogAcqBrightness
        Friend WithEvents lblNoBrightness As System.Windows.Forms.Label
        Friend WithEvents lblNoContrast As System.Windows.Forms.Label
    Private mContrast As Cognex.VisionPro.ICogAcqContrast

#Region " Windows Form Designer generated code "

    Public Sub New()
      MyBase.New()

      'This call is required by the Windows Form Designer.
      InitializeComponent()

      'Add any initialization after the InitializeComponent() call

    End Sub

    'Form overrides dispose to clean up the component list.
    Protected Overloads Overrides Sub Dispose(ByVal disposing As Boolean)
      If disposing Then
        If Not (components Is Nothing) Then
          components.Dispose()
        End If
        Dim frameGrabbers As New CogFrameGrabbers
        For Each fg As Cognex.VisionPro.ICogFrameGrabber In frameGrabbers
          fg.Disconnect(False)
        Next
      End If
      MyBase.Dispose(disposing)
    End Sub

    'Required by the Windows Form Designer
    Private components As System.ComponentModel.IContainer

    'NOTE: The following procedure is required by the Windows Form Designer
    'It can be modified using the Windows Form Designer.  
    'Do not modify it using the code editor.
    Friend WithEvents CogDisplay1 As Cognex.VisionPro.Display.CogDisplay
    Friend WithEvents txtDescription As System.Windows.Forms.TextBox
    Friend WithEvents contrastUpDown As System.Windows.Forms.NumericUpDown
    Friend WithEvents Label4 As System.Windows.Forms.Label
    Friend WithEvents brightnessUpDown As System.Windows.Forms.NumericUpDown
    Friend WithEvents Label3 As System.Windows.Forms.Label
    Friend WithEvents lblVideoFormat As System.Windows.Forms.Label
    Friend WithEvents Label2 As System.Windows.Forms.Label
    Friend WithEvents lblBoardType As System.Windows.Forms.Label
    Friend WithEvents Label1 As System.Windows.Forms.Label
    Friend WithEvents acqButton As System.Windows.Forms.Button
    <System.Diagnostics.DebuggerStepThrough()> Private Sub InitializeComponent()
            Dim resources As System.ComponentModel.ComponentResourceManager = New System.ComponentModel.ComponentResourceManager(GetType(Form1))
            Me.CogDisplay1 = New Cognex.VisionPro.Display.CogDisplay
            Me.txtDescription = New System.Windows.Forms.TextBox
            Me.contrastUpDown = New System.Windows.Forms.NumericUpDown
            Me.Label4 = New System.Windows.Forms.Label
            Me.brightnessUpDown = New System.Windows.Forms.NumericUpDown
            Me.Label3 = New System.Windows.Forms.Label
            Me.lblVideoFormat = New System.Windows.Forms.Label
            Me.Label2 = New System.Windows.Forms.Label
            Me.lblBoardType = New System.Windows.Forms.Label
            Me.Label1 = New System.Windows.Forms.Label
            Me.acqButton = New System.Windows.Forms.Button
            Me.lblNoBrightness = New System.Windows.Forms.Label
            Me.lblNoContrast = New System.Windows.Forms.Label
            CType(Me.CogDisplay1, System.ComponentModel.ISupportInitialize).BeginInit()
            CType(Me.contrastUpDown, System.ComponentModel.ISupportInitialize).BeginInit()
            CType(Me.brightnessUpDown, System.ComponentModel.ISupportInitialize).BeginInit()
            Me.SuspendLayout()
            '
            'CogDisplay1
            '
            Me.CogDisplay1.Location = New System.Drawing.Point(288, 8)
            Me.CogDisplay1.Name = "CogDisplay1"
            Me.CogDisplay1.OcxState = CType(resources.GetObject("CogDisplay1.OcxState"), System.Windows.Forms.AxHost.State)
            Me.CogDisplay1.Size = New System.Drawing.Size(328, 344)
            Me.CogDisplay1.TabIndex = 1
            '
            'txtDescription
            '
            Me.txtDescription.Location = New System.Drawing.Point(8, 360)
            Me.txtDescription.Multiline = True
            Me.txtDescription.Name = "txtDescription"
            Me.txtDescription.Size = New System.Drawing.Size(600, 32)
            Me.txtDescription.TabIndex = 2
            Me.txtDescription.Text = " This sample demonstates how to start and stop a live display using the CogDispla" & _
                "y control."
            Me.txtDescription.WordWrap = False
            '
            'contrastUpDown
            '
            Me.contrastUpDown.DecimalPlaces = 1
            Me.contrastUpDown.Increment = New Decimal(New Integer() {1, 0, 0, 65536})
            Me.contrastUpDown.Location = New System.Drawing.Point(128, 208)
            Me.contrastUpDown.Maximum = New Decimal(New Integer() {1, 0, 0, 0})
            Me.contrastUpDown.Name = "contrastUpDown"
            Me.contrastUpDown.Size = New System.Drawing.Size(89, 20)
            Me.contrastUpDown.TabIndex = 19
            '
            'Label4
            '
            Me.Label4.Location = New System.Drawing.Point(24, 208)
            Me.Label4.Name = "Label4"
            Me.Label4.Size = New System.Drawing.Size(107, 16)
            Me.Label4.TabIndex = 18
            Me.Label4.Text = "Contrast"
            '
            'brightnessUpDown
            '
            Me.brightnessUpDown.DecimalPlaces = 1
            Me.brightnessUpDown.Increment = New Decimal(New Integer() {1, 0, 0, 65536})
            Me.brightnessUpDown.Location = New System.Drawing.Point(128, 168)
            Me.brightnessUpDown.Maximum = New Decimal(New Integer() {1, 0, 0, 0})
            Me.brightnessUpDown.Name = "brightnessUpDown"
            Me.brightnessUpDown.Size = New System.Drawing.Size(89, 20)
            Me.brightnessUpDown.TabIndex = 17
            '
            'Label3
            '
            Me.Label3.Location = New System.Drawing.Point(24, 168)
            Me.Label3.Name = "Label3"
            Me.Label3.Size = New System.Drawing.Size(107, 20)
            Me.Label3.TabIndex = 16
            Me.Label3.Text = "Brightness"
            '
            'lblVideoFormat
            '
            Me.lblVideoFormat.Location = New System.Drawing.Point(8, 96)
            Me.lblVideoFormat.Name = "lblVideoFormat"
            Me.lblVideoFormat.Size = New System.Drawing.Size(264, 48)
            Me.lblVideoFormat.TabIndex = 15
            Me.lblVideoFormat.Text = "Unknown"
            '
            'Label2
            '
            Me.Label2.Location = New System.Drawing.Point(16, 72)
            Me.Label2.Name = "Label2"
            Me.Label2.Size = New System.Drawing.Size(176, 16)
            Me.Label2.TabIndex = 14
            Me.Label2.Text = "Selected Video Format"
            '
            'lblBoardType
            '
            Me.lblBoardType.Location = New System.Drawing.Point(120, 40)
            Me.lblBoardType.Name = "lblBoardType"
            Me.lblBoardType.Size = New System.Drawing.Size(152, 24)
            Me.lblBoardType.TabIndex = 13
            Me.lblBoardType.Text = "Unknown"
            '
            'Label1
            '
            Me.Label1.Location = New System.Drawing.Point(16, 40)
            Me.Label1.Name = "Label1"
            Me.Label1.Size = New System.Drawing.Size(80, 23)
            Me.Label1.TabIndex = 12
            Me.Label1.Text = "Board Type"
            '
            'acqButton
            '
            Me.acqButton.Location = New System.Drawing.Point(80, 272)
            Me.acqButton.Name = "acqButton"
            Me.acqButton.Size = New System.Drawing.Size(104, 32)
            Me.acqButton.TabIndex = 20
            Me.acqButton.Text = "Start Live Display"
            '
            'lblNoBrightness
            '
            Me.lblNoBrightness.AutoSize = True
            Me.lblNoBrightness.Location = New System.Drawing.Point(24, 170)
            Me.lblNoBrightness.Name = "lblNoBrightness"
            Me.lblNoBrightness.Size = New System.Drawing.Size(193, 13)
            Me.lblNoBrightness.TabIndex = 21
            Me.lblNoBrightness.Text = "CogAcqFifo Doesn't Support Brightness"
            Me.lblNoBrightness.Visible = False
            '
            'lblNoContrast
            '
            Me.lblNoContrast.AutoSize = True
            Me.lblNoContrast.Location = New System.Drawing.Point(24, 211)
            Me.lblNoContrast.Name = "lblNoContrast"
            Me.lblNoContrast.Size = New System.Drawing.Size(183, 13)
            Me.lblNoContrast.TabIndex = 22
            Me.lblNoContrast.Text = "CogAcqFifo Doesn't Support Contrast"
            Me.lblNoContrast.Visible = False
            '
            'Form1
            '
            Me.AutoScaleBaseSize = New System.Drawing.Size(5, 13)
            Me.ClientSize = New System.Drawing.Size(616, 398)
            Me.Controls.Add(Me.acqButton)
            Me.Controls.Add(Me.contrastUpDown)
            Me.Controls.Add(Me.Label4)
            Me.Controls.Add(Me.brightnessUpDown)
            Me.Controls.Add(Me.Label3)
            Me.Controls.Add(Me.lblVideoFormat)
            Me.Controls.Add(Me.Label2)
            Me.Controls.Add(Me.lblBoardType)
            Me.Controls.Add(Me.Label1)
            Me.Controls.Add(Me.txtDescription)
            Me.Controls.Add(Me.CogDisplay1)
            Me.Controls.Add(Me.lblNoContrast)
            Me.Controls.Add(Me.lblNoBrightness)
            Me.Name = "Form1"
            Me.Text = "Form1"
            CType(Me.CogDisplay1, System.ComponentModel.ISupportInitialize).EndInit()
            CType(Me.contrastUpDown, System.ComponentModel.ISupportInitialize).EndInit()
            CType(Me.brightnessUpDown, System.ComponentModel.ISupportInitialize).EndInit()
            Me.ResumeLayout(False)
            Me.PerformLayout()

        End Sub

#End Region
#Region " Initialization"
    Private Sub InitializeAcquisition()

      Me.Visible = True

      ' Step 1 - Create an acquisition tool which creates an CogAcqFifo with a default
      '          video format (Sony XC75 - 640 x 480).
      mTool = New CogAcqFifoTool

      ' Check if the tool was able to create a default acqfifo.
      If mTool.[Operator] Is Nothing Then
        Throw New CogAcqNoFrameGrabberException("A board might be missing or not be functioning properly.")
      End If
      ' See samples\Programming\Acquisition\Operators sample for obtaining each operator.
      ' Step 2 - Assign the CogAcqFifo, CogAcqBrightness, CogAcqContrast 
      ' First, Get the ICogAcqFifo.
      mAcqFifo = mTool.[Operator]

      ' Display the video format.
      lblVideoFormat.Text = mAcqFifo.VideoFormat
      ' Display the board type
      lblBoardType.Text = mAcqFifo.FrameGrabber.Name
      ' Let's have a small timeout period.
      mAcqFifo.Timeout = 300  ' in ms.

      ' Get the CogAcqBrightness
      ' Controls brightness levels of an acquired image
      mBrightness = mTool.[Operator].OwnedBrightnessParams
            ' Show the initial brightness if supported
            If Not mBrightness Is Nothing Then
                brightnessUpDown.Value = CType(mBrightness.Brightness, Decimal)
            Else
                brightnessUpDown.Visible = False
                lblNoBrightness.Visible = True
            End If
            ' Get the CogAcqContrast
            ' Controls contrast levels of an acquired image
            mContrast = mTool.[Operator].OwnedContrastParams
            ' Show the initial brightness if supported
            If Not mContrast Is Nothing Then
                contrastUpDown.Value = CType(mContrast.Contrast, Decimal)
            Else
                contrastUpDown.Visible = False
                lblNoContrast.Visible = True
            End If



        End Sub
#End Region
#Region " Auxiliary controls handlers"
    ' Toggle between live display and stop live display
    '
    Private Sub acqButton_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles acqButton.Click
      If CogDisplay1.LiveDisplayRunning Then
        ' Step 2 - Stop the live display
        CogDisplay1.StopLiveDisplay()
        acqButton.Text = "Start Live Display"
      Else
        ' Step 3 - Start the live display
        CogDisplay1.StartLiveDisplay(mTool.[Operator])
        acqButton.Text = "Stop Live Display"
      End If

    End Sub

    Private Sub brightnessUpDown_ValueChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles brightnessUpDown.ValueChanged

            If mBrightness Is Nothing Then Return

            If brightnessUpDown.Value >= 0 And brightnessUpDown.Value <= 1 Then
                mBrightness.Brightness = CType(brightnessUpDown.Value, Double)
            Else
                brightnessUpDown.Value = CType(0.5, Decimal)
                mBrightness.Brightness = 0.5
            End If

        End Sub
        Private Sub contrastUpDown_ValueChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles contrastUpDown.ValueChanged

            If mContrast Is Nothing Then Return

            If contrastUpDown.Value >= 0 And contrastUpDown.Value <= 1 Then
                mContrast.Contrast = CType(contrastUpDown.Value, Double)
            Else
                contrastUpDown.Value = CType(0.5, Decimal)
                mContrast.Contrast = 0.5
            End If

        End Sub
#End Region

    Private Sub Form1_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles MyBase.Load
      Try
        InitializeAcquisition()
      Catch ex As CogException
        MessageBox.Show(ex.Message)
        Application.Exit()
      Catch gex As Exception
        MessageBox.Show(gex.Message)
        Application.Exit()
      End Try

    End Sub
  End Class
End Namespace

File: Acquisition/AcqEvents/MovePart/C#/Form1.cs
/*******************************************************************************
 Copyright (C) 2004-2010 Cognex Corporation

 Subject to Cognex Corporations terms and conditions and license agreement,
 you are authorized to use and modify this source code in any way you find
 useful, provided the Software and/or the modified Software is used solely in
 conjunction with a Cognex Machine Vision System.  Furthermore you acknowledge
 and agree that Cognex has no warranty, obligations or liability for your use
 of the Software.
*******************************************************************************
 This sample program is designed to illustrate certain VisionPro features or 
 techniques in the simplest way possible. It is not intended as the framework 
 for a complete application. In particular, the sample program may not provide
 proper error handling, event handling, cleanup, repeatability, and other 
 mechanisms that a commercial quality application requires.

 This sample demonstrates how to catch and handle the move-part events.

 The move-part event fires after the camera has integrated the image, but (possibly)
 before it is available in video memory. When this event fires, it is safe to change
 the scene visible to the camera. This event is fired for all trigger mode.
 This sample will use manual triggering (also called software triggering) to
 acquire images. See StrobedAcq sample for detailed manual trigger setup procedures.

 A Cognex frame grabber must be present in order to run this sample program.
 If no Cognex board is present, the program displays an error and exits.
 
 Note that the .NET garbage collector is called on a regular basis to free
 image memory that is no longer referenced. 

 This program assumes that you have some knowledge of C# and VisionPro
 programming.

 Follow the next steps in order to catch the move-part event.
 Step 1) Create a  CogAcqFifoTool.
 Step 2) Retrieve the CogAcqMovePartEvent.
 Step 3) Display the number of times this event is called.
*/

using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using System.Data;
using System.Threading;
// Cognex namespace
using Cognex.VisionPro;
// Needed for CogException
using Cognex.VisionPro.Exceptions;

namespace MovePart
{
    /// <summary>
    /// Summary description for Form1.
    /// </summary>
    public class frmAcqEvents : System.Windows.Forms.Form
    {
        private Cognex.VisionPro.Display.CogDisplay cogDisplay1;
        private System.Windows.Forms.TextBox txtDescription;
        private System.Windows.Forms.Label label1;
        private System.Windows.Forms.Label lblBoardType;
        private System.Windows.Forms.Button acqButton;
        /// <summary>
        /// Required designer variable.
        /// </summary>
        private System.ComponentModel.Container components = null;
        private CogAcqFifoTool mTool;
        private System.Windows.Forms.Label label4;
        private System.Windows.Forms.Label lblVideoFormat;
        private System.Windows.Forms.Label label2;
        private System.Windows.Forms.Label lblMovePartCount;
        private int numAcqs = 0;
        private bool stopAcquire;
        ICogAcqFifo mAcqFifo;

        public frmAcqEvents()
        {
            //
            // Required for Windows Form Designer support
            //
            InitializeComponent();
            InitializeAcquisition();

            //
            // TODO: Add any constructor code after InitializeComponent call
            //
        }


    protected override void OnClosing(CancelEventArgs e)
    {
      stopAcquire = true;
      if (mAcqFifo != null)
      { 
          mAcqFifo.Flush();
      }
      int counter = 0;
      while ( counter < 10)
      {
        Application.DoEvents();
        Thread.Sleep(1);
        counter++;
      }
      base.OnClosing (e);
      
    }

        /// <summary>
        /// Clean up any resources being used.
        /// </summary>
        protected override void Dispose( bool disposing )
        {
            if( disposing )
            {
        if (components != null) 
                {
                    components.Dispose();
                }
        CogFrameGrabbers frameGrabbers = new CogFrameGrabbers();
        foreach (ICogFrameGrabber fg in frameGrabbers)
          fg.Disconnect(false);
      }
            base.Dispose( disposing );
        }

        #region Windows Form Designer generated code
        /// <summary>
        /// Required method for Designer support - do not modify
        /// the contents of this method with the code editor.
        /// </summary>
        private void InitializeComponent()
        {
            System.Resources.ResourceManager resources = new System.Resources.ResourceManager(typeof(frmAcqEvents));
            this.cogDisplay1 = new Cognex.VisionPro.Display.CogDisplay();
            this.txtDescription = new System.Windows.Forms.TextBox();
            this.label1 = new System.Windows.Forms.Label();
            this.lblBoardType = new System.Windows.Forms.Label();
            this.acqButton = new System.Windows.Forms.Button();
            this.label4 = new System.Windows.Forms.Label();
            this.lblVideoFormat = new System.Windows.Forms.Label();
            this.label2 = new System.Windows.Forms.Label();
            this.lblMovePartCount = new System.Windows.Forms.Label();
            ((System.ComponentModel.ISupportInitialize)(this.cogDisplay1)).BeginInit();
            this.SuspendLayout();
            // 
            // cogDisplay1
            // 
            this.cogDisplay1.Location = new System.Drawing.Point(232, 24);
            this.cogDisplay1.Name = "cogDisplay1";
            this.cogDisplay1.OcxState = ((System.Windows.Forms.AxHost.State)(resources.GetObject("cogDisplay1.OcxState")));
            this.cogDisplay1.Size = new System.Drawing.Size(304, 296);
            this.cogDisplay1.TabIndex = 0;
            // 
            // txtDescription
            // 
            this.txtDescription.Location = new System.Drawing.Point(8, 328);
            this.txtDescription.Multiline = true;
            this.txtDescription.Name = "txtDescription";
            this.txtDescription.ReadOnly = true;
            this.txtDescription.Size = new System.Drawing.Size(528, 48);
            this.txtDescription.TabIndex = 1;
            this.txtDescription.Text = @"This sample demonstrates how to handle the move-part event.  A Cognex frame grabber board must be present in order to run this sample program. When the Run button is pressed, the program first flushes all outstanding acquisitions since they are not part of new acquisitions, and acquires images automatically.";
            // 
            // label1
            // 
            this.label1.Location = new System.Drawing.Point(16, 32);
            this.label1.Name = "label1";
            this.label1.Size = new System.Drawing.Size(72, 16);
            this.label1.TabIndex = 2;
            this.label1.Text = "Board Type:";
            // 
            // lblBoardType
            // 
            this.lblBoardType.Location = new System.Drawing.Point(96, 32);
            this.lblBoardType.Name = "lblBoardType";
            this.lblBoardType.Size = new System.Drawing.Size(128, 16);
            this.lblBoardType.TabIndex = 3;
            this.lblBoardType.Text = "Unknown";
            // 
            // acqButton
            // 
            this.acqButton.Location = new System.Drawing.Point(40, 128);
            this.acqButton.Name = "acqButton";
            this.acqButton.Size = new System.Drawing.Size(104, 32);
            this.acqButton.TabIndex = 10;
            this.acqButton.Text = "Run";
            this.acqButton.Click += new System.EventHandler(this.acqButton_Click);
            // 
            // label4
            // 
            this.label4.Location = new System.Drawing.Point(16, 56);
            this.label4.Name = "label4";
            this.label4.Size = new System.Drawing.Size(128, 16);
            this.label4.TabIndex = 11;
            this.label4.Text = "Selected Video Format:";
            // 
            // lblVideoFormat
            // 
            this.lblVideoFormat.Location = new System.Drawing.Point(16, 72);
            this.lblVideoFormat.Name = "lblVideoFormat";
            this.lblVideoFormat.Size = new System.Drawing.Size(208, 16);
            this.lblVideoFormat.TabIndex = 12;
            this.lblVideoFormat.Text = "Unknown";
            // 
            // label2
            // 
            this.label2.Location = new System.Drawing.Point(32, 200);
            this.label2.Name = "label2";
            this.label2.Size = new System.Drawing.Size(120, 16);
            this.label2.TabIndex = 13;
            this.label2.Text = "Move part event count:";
            // 
            // lblMovePartCount
            // 
            this.lblMovePartCount.Location = new System.Drawing.Point(160, 200);
            this.lblMovePartCount.Name = "lblMovePartCount";
            this.lblMovePartCount.Size = new System.Drawing.Size(64, 16);
            this.lblMovePartCount.TabIndex = 14;
            this.lblMovePartCount.Text = "0";
            // 
            // frmAcqEvents
            // 
            this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
            this.ClientSize = new System.Drawing.Size(536, 390);
            this.Controls.Add(this.lblMovePartCount);
            this.Controls.Add(this.label2);
            this.Controls.Add(this.lblVideoFormat);
            this.Controls.Add(this.label4);
            this.Controls.Add(this.acqButton);
            this.Controls.Add(this.lblBoardType);
            this.Controls.Add(this.label1);
            this.Controls.Add(this.txtDescription);
            this.Controls.Add(this.cogDisplay1);
            this.Name = "frmAcqEvents";
            this.Text = "Show how to capture the move part event";
            ((System.ComponentModel.ISupportInitialize)(this.cogDisplay1)).EndInit();
            this.ResumeLayout(false);

        }
        #endregion

        /// <summary>
        /// The main entry point for the application.
        /// </summary>
        [STAThread]
        static void Main() 
        {
            try 
            {
                Application.Run(new frmAcqEvents());
            }
            catch (CogException ce)
            {	
                MessageBox.Show("The following error has occured\n" + ce.Message);
                Application.Exit();
                
            }
                
        }

        public void InitializeAcquisition()
        {

            try
            {
                // Step 1 - Create an acquisition tool which creates an CogAcqFifo with a default
                //           video format for that frame grabber.
                mTool = new CogAcqFifoTool();

                // Check if the tool was able to create a default acqfifo.
                if (mTool.Operator == null)
                    throw new CogAcqNoFrameGrabberException("A board might be missing or not be functioning properly.");

                // First, Get the ICogAcqFifo.
                mAcqFifo = mTool.Operator;
                // Display the video format
                lblVideoFormat.Text = mAcqFifo.VideoFormat;
                // Display the board type
                lblBoardType.Text = mAcqFifo.FrameGrabber.Name;

                // Hook up the acquisition completion event. Each time a tool acquires
                // an image, it fires the acquisition completion event.
                mAcqFifo.Complete += new CogCompleteEventHandler(Operator_Complete);

                // Hook up the acquisition move part event. 
                mAcqFifo.MovePart += new CogMovePartEventHandler(Operator_MovePart);

                // NOTE: Either the exposure or brightness may need adjustment to clearly see
                // the acquired image. Both exposure and brightness are set to high values
                // in case sufficient lighting is unavailable.
                if (mAcqFifo.OwnedExposureParams != null)
                    mAcqFifo.OwnedExposureParams.Exposure = 50; //mSecs
                if (mAcqFifo.OwnedBrightnessParams != null)
                    mAcqFifo.OwnedBrightnessParams.Brightness = 0.9;
            }
            catch (CogAcqException ex)
            {
                MessageBox.Show("The following error has occured\n" + ex.Message);             
            }
        }

        // This is the complete event handler for acquisition.  When an image is acquired,
        // it fires a complete event.  This handler verifies the state of the acquisition
        // fifo, and then calls Complete(), which gets the image from the fifo.
        
        // Note that it is necessary to call the .NET garbarge collector on a regular
        // basis so large images that are no longer used will be released back to the
        // heap.  In this sample, it is called every 5th acqusition.
        private void Operator_Complete(object sender, CogCompleteEventArgs e)
        {
            if (InvokeRequired)
            {
                Invoke(new CogCompleteEventHandler(Operator_Complete),
                    new object[] {sender, e});
                return;
            }

            int numReadyVal, numPendingVal;
            bool busyVal;
            CogAcqInfo info = new CogAcqInfo();
            try
            {
                // When you click the Stop button, the program calls the Flush method. However, the
                // acquisition completion event might have been fired prior to calling Flush. If this
                // happens mAcqFifo.Complete() will fail. To prevent this problem,
                // the sample uses the StopAcquire boolean flag.
                if (stopAcquire)
                    return;

                mAcqFifo.GetFifoState(out numPendingVal,out numReadyVal,out busyVal);
                if (numReadyVal > 0 )
                    cogDisplay1.Image = 
                        mAcqFifo.CompleteAcquireEx(info); 
                numAcqs++;
                // We need to run the garbage collector on occasion to cleanup
                // images that are no longer being used.
                if (numAcqs > 4)
                {
                    GC.Collect();
                    numAcqs = 0;
                }
                mAcqFifo.StartAcquire();  // request another acquisition
            }
            catch (CogException ce)
            {
                MessageBox.Show("The following error has occured\n" + ce.Message);
            }
        }

        private void Operator_MovePart(object sender, CogMovePartEventArgs e)
        {
            // this next bit of code converts a string to an int32, adds 1,
            // and then converts it back to a string for displaying as a label
                    if (InvokeRequired)
                    {
                        Invoke(new CogMovePartEventHandler(Operator_MovePart),
                            new object[] {sender, e});
                        return;
                    }

            lblMovePartCount.Text = ((System.Int32)(System.Int32.Parse(lblMovePartCount.Text,
                System.Globalization.NumberStyles.Integer,null) + 1)).ToString();
        }

        // This method is changed event handler for the acqButton control
        // on form.  It starts image acquisition.
        private void acqButton_Click(object sender, System.EventArgs e)
        {
            try
            {
                if (mAcqFifo == null)
                    throw new CogAcqNoFrameGrabberException("A board might be missing or not be functioning properly. Can not run");
                else
                {
                    if (acqButton.Text == "Run")
                    {
                        stopAcquire = false;
                        lblMovePartCount.Text = "0";
                        // Flush all outstanding acquisitions since they are not part of new acquisitions.
                        mAcqFifo.Flush();
                        // Call StartAcquire to issue an acquisition request and wait for
                        // the acquisition complete event
                        // Note: For manual triggering, prime the acquisition engine
                        // with two (or more, up to 32) start requests to ensure optimal throughput.
                        mAcqFifo.StartAcquire();
                        mAcqFifo.StartAcquire();
                        acqButton.Text = "Stop";
                    }
                    else
                    {
                        stopAcquire = true;
                        // Flush all outstanding acquisition requests and stop.
                        mAcqFifo.Flush();
                        acqButton.Text = "Run";
                    }
                }

            }
            catch (CogException ce)
            {
                MessageBox.Show("The following error has occured\n" + ce.Message);
            }
        }

    }
}

